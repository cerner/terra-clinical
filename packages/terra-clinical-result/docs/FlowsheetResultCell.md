# Terra Clinical Flowsheet Result Cell

The Clinical Result is a standardized view for single result values added to a Patient's Medical Chart. The flowsheet result cell is specifically for the grid display of a result collection, and can handle displaying an indication of multiple results.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
```
$ npm install terra-clinical-result
```

<!-- AUTO-GENERATED-CONTENT:START Peer Dependencies -->
<!-- AUTO-GENERATED-CONTENT:END -->

## Component Features

* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
* [LTR/RTL Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#ltr--rtl-support)

## Usage Guide

The flowsheet cell for clinical results is constructed by providing the same structured object as the clinical result in an array, allowing for multiple results to be indicated in the same time cell.

Currently the Flowsheet expects an array of one or more results - and if there are multiple, it will display the first in the array, with an indication of additional results. Updates are in progress to also look at the performedDateTime value if provided to evaluate which result is the most recent to appropriately display, regardless of which order index position in the array it is.


```jsx
/* ------ Structure for the resultData object for the Flowsheet Result Cell, can support multiple results ------ */

 /**
  *  A set of clinical results.                                                          .
  */

  resultDataSet = [
    {
      /**
       *  A single clinical result or blood pressure result.
       *  Same as {resultData} for clinical-result and clinical-result-blood-pressure
       */
      eventId:'1574537491',
      result: { /* ... */ },
    },
    {
      /**
       *  A single clinical result or blood pressure result.
       *  Same as {resultData} for clinical-result and clinical-result-blood-pressure
       */
      eventId:'1574537491',
      result: { /* ... */ },
    },
    {
      /**
       *  A single clinical result or blood pressure result.
       *  Same as {resultData} for clinical-result and clinical-result-blood-pressure
       */
      eventId:'1574537491',
      result: { /* ... */ },
    },
  ]

```


An example of a single clinical result value in a flowsheet cell:
```jsx
import React from 'react';
import { FlowsheetResultCell } from 'terra-clinical-result/lib/index';

const singleResultValue = [
  {
    eventId: '1574537491',
    result: {
      value: '101.9',
      unit: 'degC',
    },
    interpretation: 'critical',
    isNumeric: true,
    isModified: true,
    hasComment: true,
    isUnverified: true,
    conceptDisplay: 'Temperature Oral',
    datetimeDisplay: 'Nov 23, 2019 13:31:31',
  },
];

export default () => <FlowsheetResultCell resultDataSet={singleResultValue} />;

```


An example of multiple clinical blood pressure result values:
```jsx
import React from 'react';
import { FlowsheetResultCell } from 'terra-clinical-result/lib/index';
import classNames from 'classnames/bind';
import styles from '../Examples.module.scss';

const cx = classNames.bind(styles);

const multipleResultBPValues = [
  {
    id: '1577895000-1',
    systolic: {
      eventId: '1577895000-1.1',
      result: {
        value: '111',
        unit: 'mmHg',
      },
      interpretation: 'high',
      isModified: true,
      hasComment: true,
    },
    diastolic: {
      eventId: '1577895000-1.2',
      result: {
        value: '77',
        unit: 'mmHg',
      },
      interpretation: 'low',
      isModified: true,
      hasComment: true,
    },
  },
  {
    id: '1577895000-2',
    systolic: {
      eventId: '1577895000-2.1',
      result: {
        value: '111',
        unit: 'mmHg',
      },
      interpretation: 'critical',
      isModified: true,
      hasComment: true,
    },
    diastolic: {
      eventId: '1577895000-2.2',
      result: {
        value: '77',
        unit: 'mmHg',
      },
      isModified: true,
      hasComment: true,
    },
  },
  {
    id: '1577895000-3',
    systolic: {
      eventId: '1577895000-3.1',
      result: {
        value: '111',
        unit: 'mmHg',
      },
      interpretation: 'critical',
      isModified: true,
      hasComment: true,
    },
    diastolic: {
      eventId: '1577895000-3.2',
      result: {
        value: '77',
        unit: 'mmHg',
      },
      isModified: true,
      hasComment: true,
    },
  },
];

export default () => <FlowsheetResultCell resultDataSet={multipleResultBPValues} hideUnit />;

```
