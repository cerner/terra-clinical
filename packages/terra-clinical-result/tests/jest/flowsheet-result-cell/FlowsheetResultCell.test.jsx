import React from 'react';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { shallowWithIntl } from 'terra-enzyme-intl';
import FlowsheetResultCell from '../../../src/flowsheet-result-cell/FlowsheetResultCell';
import DefaultResult, {
  DefaultBloodPressureResult, DefaultSystolicResult, DefaultDiastolicResult, DefaultResultWithNoEventId, DefaultBloodPressureDiastolicResultWithNoId, DefaultBloodPressureSystolicResultWithNoId,
} from '../../../src/terra-dev-site/test/clinical-result/TestResults';

// Snapshot Tests
describe('FlowsheetResultCell', () => {
  it('should render a ResultError if hasResultError is true', () => {
    const result = shallowWithIntl(<FlowsheetResultCell hasResultError />).dive();
    expect(result).toMatchSnapshot();
  });

  it('should render a NoData if hasResultNoData is true', () => {
    const result = shallowWithIntl(<FlowsheetResultCell hasResultNoData />).dive();
    expect(result).toMatchSnapshot();
  });

  it('should render an error with bad data', () => {
    const result = shallowWithIntl(<FlowsheetResultCell />).dive();
    expect(result).toMatchSnapshot();
  });

  it('should render when given result data', () => {
    const results = [
      DefaultResult,
      DefaultBloodPressureResult,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render when given result data with no event id', () => {
    const results = [
      DefaultResultWithNoEventId,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render when given blood pressure systolic result data with no id', () => {
    const results = [
      DefaultBloodPressureSystolicResultWithNoId,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render when given blood pressure diastolic result data with no id', () => {
    const results = [
      DefaultBloodPressureDiastolicResultWithNoId,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render an error when given no result data', () => {
    const cell = shallowWithIntl(<FlowsheetResultCell />);
    expect(cell).toMatchSnapshot();
  });

  it('should pass hideUnit down', () => {
    const results = [
      DefaultResult,
      DefaultBloodPressureResult,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} hideUnit />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render a entered-in-error status standard result', () => {
    const results = [
      {
        ...DefaultResult,
        status: 'entered-in-error',
        interpretation: 'critical',
      },
      DefaultResult,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} hideUnit />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render correctly with an interpretation and unverified result', () => {
    const results = [
      {
        ...DefaultResult,
        isUnverified: true,
        interpretation: 'critical',
      },
      {
        ...DefaultResult,
        isUnverified: true,
        interpretation: 'critical',
      },
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} hideUnit />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render systolic entered-in-error status bloodpressure result', () => {
    const results = [
      {
        id: '111',
        systolic: {
          ...DefaultSystolicResult,
          status: 'entered-in-error',
          interpretation: 'critical',
        },
        diastolic: {
          ...DefaultDiastolicResult,
        },
      },
      DefaultBloodPressureResult,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} hideUnit />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render both entered-in-error status bloodpressure result', () => {
    const results = [
      {
        id: '111',
        systolic: {
          ...DefaultSystolicResult,
          status: 'entered-in-error',
          interpretation: 'critical',
        },
        diastolic: {
          ...DefaultDiastolicResult,
          status: 'entered-in-error',
          interpretation: 'critical',
        },
      },
      DefaultBloodPressureResult,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} hideUnit />).dive();
    expect(cell).toMatchSnapshot();
  });

  it('should render the extra results icon with no regard to an entered in error result\'s interpretation.', () => {
    const results = [
      DefaultResult,
      {
        id: '111',
        systolic: {
          ...DefaultSystolicResult,
          status: 'entered-in-error',
          interpretation: 'critical',
        },
        diastolic: {
          ...DefaultDiastolicResult,
          status: 'entered-in-error',
          interpretation: 'critical',
        },
      },
      DefaultBloodPressureResult,
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} hideUnit />).dive();
    expect(cell.find('.additional-end-display').hasClass('interpretation-critical')).toBe(false);
  });

  it('should render blood pressure correctly with an interpretation and unverified result', () => {
    const results = [
      {
        id: '111',
        systolic: {
          ...DefaultSystolicResult,
          isUnverified: true,
          interpretation: 'critical',
        },
        diastolic: {
          ...DefaultDiastolicResult,
          isUnverified: true,
          interpretation: 'critical',
        },
      },
      {
        id: '111',
        systolic: {
          ...DefaultSystolicResult,
          isUnverified: true,
          interpretation: 'critical',
        },
        diastolic: {
          ...DefaultDiastolicResult,
          isUnverified: true,
          interpretation: 'critical',
        },
      },
    ];
    const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} />).dive();
    expect(cell).toMatchSnapshot();
  });

  describe('paddingStyle', () => {
    it('should render when given none', () => {
      const results = [
        DefaultResult,
        DefaultBloodPressureResult,
      ];
      const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} paddingStyle="none" />).dive();
      expect(cell).toMatchSnapshot();
    });

    it('should render when given standard', () => {
      const results = [
        DefaultResult,
        DefaultBloodPressureResult,
      ];
      const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} paddingStyle="standard" />).dive();
      expect(cell).toMatchSnapshot();
    });

    it('should render when given compact', () => {
      const results = [
        DefaultResult,
        DefaultBloodPressureResult,
      ];
      const cell = shallowWithIntl(<FlowsheetResultCell resultDataSet={results} paddingStyle="compact" />).dive();
      expect(cell).toMatchSnapshot();
    });
  });
});
