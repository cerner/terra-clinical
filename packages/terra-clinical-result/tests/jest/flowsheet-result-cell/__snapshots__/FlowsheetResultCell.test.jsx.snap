// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowsheetResultCell paddingStyle should render when given compact 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      eventId="111"
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell paddingStyle should render when given none 1`] = `
<div
  className="flowsheet-result-cell"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      eventId="111"
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell paddingStyle should render when given standard 1`] = `
<div
  className="flowsheet-result-cell padding-standard"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      eventId="111"
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should pass hideUnit down 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      eventId="111"
      hideAccessoryDisplays={true}
      hideUnit={true}
      isTruncated={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render a NoData if hasResultNoData is true 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-NoData"
  >
    <InjectIntl(KnownNoData) />
  </div>
</div>
`;

exports[`FlowsheetResultCell should render a ResultError if hasResultError is true 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display error"
    key="ClinicalResultDisplay-Error"
  >
    <InjectIntl(ResultError) />
  </div>
</div>
`;

exports[`FlowsheetResultCell should render a entered-in-error status standard result 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display interpretation"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(EnteredInError) />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render an error when given no result data 1`] = `
<FlowsheetResultCell
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  paddingStyle="compact"
  resultDataSet={Array []}
/>
`;

exports[`FlowsheetResultCell should render an error with bad data 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display error"
    key="ClinicalResultDisplay-Error"
  >
    <InjectIntl(ResultError) />
  </div>
</div>
`;

exports[`FlowsheetResultCell should render blood pressure correctly with an interpretation and unverified result 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResultBloodPressure)
      diastolic={
        Object {
          "eventId": "111.2",
          "interpretation": "critical",
          "isUnverified": true,
          "result": Object {
            "unit": "mmHg",
            "value": "80",
          },
        }
      }
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResultBloodPressure-111"
      systolic={
        Object {
          "eventId": "111.1",
          "interpretation": "critical",
          "isUnverified": true,
          "result": Object {
            "unit": "mmHg",
            "value": "180",
          },
        }
      }
    />
  </div>
  <div
    className="additional-end-display"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
  <div
    className="end-accessory-icons"
    key="EndAccessoryIcons-111"
  >
    <div
      className="end-accessory-stack"
    >
      <IconDiamond
        className="icon-unverified"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render both entered-in-error status bloodpressure result 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display interpretation"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(EnteredInError) />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render correctly with an interpretation and unverified result 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      eventId="111"
      hideAccessoryDisplays={true}
      hideUnit={true}
      interpretation="critical"
      isTruncated={true}
      isUnverified={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
  <div
    className="additional-end-display"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
  <div
    className="end-accessory-icons"
    key="EndAccessoryIcons-111"
  >
    <div
      className="end-accessory-stack"
    >
      <IconDiamond
        className="icon-unverified"
        viewBox="0 0 48 48"
        xmlns="http://www.w3.org/2000/svg"
      />
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render systolic entered-in-error status bloodpressure result 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display interpretation"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(EnteredInError) />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render when given blood pressure diastolic result data with no id 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111.2"
  >
    <InjectIntl(ClinicalResultBloodPressure)
      diastolic={
        Object {
          "eventId": "111.2",
          "result": Object {
            "unit": "mmHg",
            "value": "80",
          },
        }
      }
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResultBloodPressure-111.2"
    />
  </div>
</div>
`;

exports[`FlowsheetResultCell should render when given blood pressure systolic result data with no id 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111.1"
  >
    <InjectIntl(ClinicalResultBloodPressure)
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResultBloodPressure-111.1"
      systolic={
        Object {
          "eventId": "111.1",
          "result": Object {
            "unit": "mmHg",
            "value": "180",
          },
        }
      }
    />
  </div>
</div>
`;

exports[`FlowsheetResultCell should render when given result data 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      eventId="111"
      hideAccessoryDisplays={true}
      isTruncated={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
  <div
    className="additional-end-display no-accessory-icons"
    key="AdditionalResultsDisplay-111"
  >
    <div
      className="additional-results-stack"
    >
      <span
        className="additional-results-value"
      >
        2
      </span>
      <span
        className="additional-results-value"
      >
        2
      </span>
    </div>
  </div>
</div>
`;

exports[`FlowsheetResultCell should render when given result data with no event id 1`] = `
<div
  className="flowsheet-result-cell padding-compact"
>
  <div
    className="primary-display"
    key="ClinicalResultDisplay-111"
  >
    <InjectIntl(ClinicalResult)
      hideAccessoryDisplays={true}
      id="111"
      isTruncated={true}
      key="ClinicalResult-111"
      result={
        Object {
          "unit": "mL",
          "value": "12345.678",
        }
      }
    />
  </div>
</div>
`;
